File: examples/cpp/state.cpp
#include <iostream>
#include <memory>
#include <string>

class Document;

class State {
public:
    virtual ~State() = default;
    virtual void publish(Document& doc) = 0;
    virtual std::string name() const = 0;
};

class Draft : public State {
public:
    void publish(Document& doc) override;
    std::string name() const override { return "Draft"; }
};

class Review : public State {
public:
    void publish(Document& doc) override;
    std::string name() const override { return "Review"; }
};

class Published : public State {
public:
    void publish(Document& doc) override {
        std::cout << "  Already published. No changes." << std::endl;
    }
    std::string name() const override { return "Published"; }
};

class Document {
    std::unique_ptr<State> state_;
    std::string content_;
public:
    Document(const std::string& content) : state_(std::make_unique<Draft>()), content_(content) {}
    void setState(std::unique_ptr<State> state) { state_ = std::move(state); }
    void publish() {
        std::cout << "  Current state: " << state_->name() << std::endl;
        state_->publish(*this);
    }
    std::string getStateName() const { return state_->name(); }
};

void Draft::publish(Document& doc) {
    std::cout << "  Moving from Draft to Review" << std::endl;
    doc.setState(std::make_unique<Review>());
}

void Review::publish(Document& doc) {
    std::cout << "  Moving from Review to Published" << std::endl;
    doc.setState(std::make_unique<Published>());
}

int main() {
    std::cout << "=== State: Document Workflow ===" << std::endl;

    Document doc("My Article");

    std::cout << "\nPublish attempt 1:" << std::endl;
    doc.publish();

    std::cout << "\nPublish attempt 2:" << std::endl;
    doc.publish();

    std::cout << "\nPublish attempt 3:" << std::endl;
    doc.publish();

    std::cout << "\nFinal state: " << doc.getStateName() << std::endl;

    return 0;
}

File: examples/delphi/state.pas
{$mode objfpc}{$H+}
program state;

uses SysUtils;

type
  TDocument = class;

  TDocState = class(TObject)
  public
    function GetName: string; virtual; abstract;
    procedure Publish(Doc: TDocument); virtual; abstract;
  end;

  TDraftState = class(TDocState)
  public
    function GetName: string; override;
    procedure Publish(Doc: TDocument); override;
  end;

  TReviewState = class(TDocState)
  public
    function GetName: string; override;
    procedure Publish(Doc: TDocument); override;
  end;

  TPublishedState = class(TDocState)
  public
    function GetName: string; override;
    procedure Publish(Doc: TDocument); override;
  end;

  TDocument = class(TObject)
  private
    FState: TDocState;
    FTitle: string;
  public
    constructor Create(const ATitle: string);
    destructor Destroy; override;
    procedure ChangeState(NewState: TDocState);
    procedure Publish;
    procedure Show;
    property Title: string read FTitle;
  end;

function TDraftState.GetName: string;
begin
  Result := 'Draft';
end;

procedure TDraftState.Publish(Doc: TDocument);
begin
  WriteLn('  Moving from Draft to Review');
  Doc.ChangeState(TReviewState.Create);
end;

function TReviewState.GetName: string;
begin
  Result := 'Review';
end;

procedure TReviewState.Publish(Doc: TDocument);
begin
  WriteLn('  Moving from Review to Published');
  Doc.ChangeState(TPublishedState.Create);
end;

function TPublishedState.GetName: string;
begin
  Result := 'Published';
end;

procedure TPublishedState.Publish(Doc: TDocument);
begin
  WriteLn('  Already published, no change');
end;

constructor TDocument.Create(const ATitle: string);
begin
  inherited Create;
  FTitle := ATitle;
  FState := TDraftState.Create;
end;

destructor TDocument.Destroy;
begin
  FState.Free;
  inherited;
end;

procedure TDocument.ChangeState(NewState: TDocState);
begin
  FState.Free;
  FState := NewState;
end;

procedure TDocument.Publish;
begin
  FState.Publish(Self);
end;

procedure TDocument.Show;
begin
  WriteLn('  Document "', FTitle, '" is in state: ', FState.GetName);
end;

var
  Doc: TDocument;
begin
  WriteLn('=== State Pattern: Document Workflow ===');
  WriteLn;

  Doc := TDocument.Create('Design Patterns Guide');
  Doc.Show;
  WriteLn;

  Doc.Publish;
  Doc.Show;
  WriteLn;

  Doc.Publish;
  Doc.Show;
  WriteLn;

  Doc.Publish;
  Doc.Show;

  Doc.Free;
end.
